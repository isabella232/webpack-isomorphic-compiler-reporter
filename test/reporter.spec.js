'use strict';

const path = require('path');
const createReporter = require('../');
const createCompiler = require('./util/createCompiler');
const createStats = require('./util/createStats');
const createWritter = require('./util/createWritter');

const dateNow = 1513447567843;

beforeEach(() => jest.spyOn(Date, 'now').mockImplementation(() => 1513447567843));
afterEach(() => jest.restoreAllMocks());

it('should render the correct output on success', () => {
    const compiler = createCompiler();
    const writter = createWritter();
    const stats = {
        client: createStats(),
        server: createStats(),
    };

    createReporter(compiler, { write: writter });

    compiler.emit('begin');
    compiler.emit('end', stats);

    expect(writter.getOutput()).toMatchSnapshot();
});

it('should render the correct output on failure', () => {
    const compiler = createCompiler();
    const writter = createWritter();
    const stats = createStats(true);
    const error = Object.assign(new Error('Error message'), {
        stats,
        stack: [
            'at method1 (/path/to/file1.js:1:0)',
            'at method2 (/path/to/file2.js:1:0)',
        ].join('\n'),
    });

    createReporter(compiler, { stats: 'once', write: writter });

    compiler.emit('begin');
    compiler.emit('error', error);

    expect(writter.getOutput()).toMatchSnapshot();
});

it('should calculate the correct duration', () => {
    const compiler = createCompiler();
    const writter = createWritter();
    const stats = {
        client: createStats(),
        server: createStats(),
    };

    createReporter(compiler, { write: writter });

    compiler.emit('begin');
    compiler.emit('end', stats);

    Date.now.mockImplementation(() => dateNow + 100);
    compiler.emit('begin');
    Date.now.mockImplementation(() => dateNow + 300);
    compiler.emit('end', stats);

    expect(writter.getOutput()).toMatchSnapshot();
});

describe('human errors', () => {
    it('should warn about all human errors', () => {
        const badServerConfig = {
            entry: path.resolve(`${__dirname}/files/simple.js`),
            output: {
                path: path.resolve(`${__dirname}/../tmp`),
                filename: 'client.js',
            },
            devtool: 'eval-source-map',
        };
        const compiler = createCompiler({
            server: { webpackConfig: badServerConfig },
        });
        const writter = createWritter();

        createReporter(compiler, { write: writter });

        compiler.emit('begin');

        expect(writter.getOutput()).toMatchSnapshot();
    });

    it('should not warn about human errors if `options.humanErrors` is false', () => {
        const badServerConfig = {
            entry: path.resolve(`${__dirname}/files/simple.js`),
            output: {
                path: path.resolve(`${__dirname}/../tmp`),
                filename: 'client.js',
            },
            devtool: 'eval-source-map',
        };
        const compiler = createCompiler({
            server: { webpackConfig: badServerConfig },
        });
        const writter = createWritter();

        createReporter(compiler, {
            write: writter,
            humanErrors: false,
        });

        compiler.emit('begin');

        expect(writter.getOutput()).toMatchSnapshot();
    });

    it('should stop reporting if stop was called', () => {
        const compiler = createCompiler();
        const writter = createWritter();
        const stats = {
            client: createStats(),
            server: createStats(),
        };

        const { stop } = createReporter(compiler, { write: writter });

        compiler.emit('begin');
        compiler.emit('end', stats);

        stop();

        compiler.emit('begin');
        compiler.emit('end', stats);

        expect(writter.getOutput()).toMatchSnapshot();
    });
});
