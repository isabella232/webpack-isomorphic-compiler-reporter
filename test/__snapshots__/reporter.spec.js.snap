// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`human errors should not warn about human errors if \`options.humanErrors\` is false 1`] = `
"● Compiling...
"
`;

exports[`human errors should stop reporting if stop was called 1`] = `
"● Compiling...
✔ Compilation succeeded (0ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    foo.js   10Kb

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    bar.js   10Kb

"
`;

exports[`human errors should warn about all human errors 1`] = `
"● Compiling...
WARN: Server \`output.target\` in webpack config should be set to \`node\`
See https://webpack.js.org/configuration/target/#target

WARN: Server \`output.libraryTarget\` in webpack config should be set to \`this\` or \`commonjs2\`
See https://webpack.js.org/configuration/output/#output-librarytarget

WARN: Server \`devtool\` in webpack config should be set to \`source-map\` or \`inline-source-map\`
In addition, you must use https://github.com/evanw/node-source-map-support to enable source maps.

WARN: The \`output.path\` should be the same in the webpack client & server configs
Client output.path: /build
Server output.path: /foo

WARN: The \`output.publicPath\` should be the same in the webpack client & server configs
Client output.publicPath: /build/
Server output.publicPath: /foo/

"
`;

exports[`should output the correct duration 1`] = `
"● Compiling...
✔ Compilation succeeded (0ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    foo.js   10Kb

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    bar.js   10Kb

● Compiling...
✔ Compilation succeeded (200ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    foo.js   10Kb

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    bar.js   10Kb

"
`;

exports[`should render the correct output on failure 1`] = `
"● Compiling...
✖ Compilation failed

    Error message
    - file1.js:1 method1
        /path/to/file1.js:1:0
      - file2.js:1 method2
        /path/to/file2.js:1:0

"
`;

exports[`should render the correct output on success 1`] = `
"● Compiling...
✔ Compilation succeeded (0ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    foo.js   10Kb

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset    Size
    bar.js   10Kb

"
`;
